### PR TITLE
filled in router methods

### PR DESCRIPTION
**Pull request by malavika1**
also, made links a dict instead of an array

### CODE REVIEW

**Comment by malavika1** [comment]
@mortonrj routing packets need a destination, link, and cost eventually

File: router.py [hunk 1]
```diff
@@ -44,8 +56,43 @@ def receive_packet(self, packet):
...
 
-    def make_routing_table(self):
-        """Makes the routing table based on the links."""
-        pass
\ No newline at end of file
+        # If we get a routing packet, we need to update the routing table
+        if isinstance(packet, RoutingPacket):
+            self.update_routing_table(packet, link_id)
+        else: # We need to send the packet to its dest host
+            self.send_packet(packet)
```
Comment (mortonrj) [inline router.py]: Is this right?  Is self the source?  (it probs is, I'm just confused cause the flow life is the only one I know)

File: router.py [hunk 2]
```diff
@@ -44,8 +56,43 @@ def receive_packet(self, packet):
...
+
+    def send_routing_packets(self):
+        """Sends out the routing table to all of the router's links."""
+
+        for _, link in links:
```
Comment (mortonrj) [inline router.py]: Should this be self.links instead of links?

File: router.py [hunk 3]
```diff
@@ -44,8 +56,43 @@ def receive_packet(self, packet):
...
+            src = self.node_id
+            dest = link.get_other_node_id(self.node_id)
+            packet = RoutingPacket(None, src, dest, None, self.routing_table)
+            event = lambda: link.add_packet(packet, self.node_id)
+            ns.add_event(event)
```
Comment (mortonrj) [inline router.py]: Should this be self.ns.add_event(event)?

File: router.py (hunk 1 above)
Comment (malavika1) [inline router.py]: i'm calling the router object's send_packet method so it should be right. idk what you mean by is self the source

Comment (Jdonerme) [inline router.py]: @malavika1 is right I'm p sure. Even though other packets are generated by flows, routing table packets are generated by routers and should just be created by the router

File: router.py [hunk 4]
```diff
@@ -1,5 +1,6 @@
 from node import Node
 from routingpacket import RoutingPacket
+from constants import *
```
Comment (Jdonerme) [inline router.py]: Not using?

File: router.py [hunk 5]
```diff
@@ -32,9 +38,14 @@ def send_packet(self, packet):
             packet (Packet): The packet to send.
 
         """
-        pass
 
-    def receive_packet(self, packet):
+        # Use the routing table to send a packet to the right node
+        link, _ = self.routing_table[packet.dest.node_id]
```
Comment (hatooku) [inline router.py]: self.routing_table[packet.dest.node_id][0] is cleaner

File: router.py (hunk 1 above)
Comment (hatooku) [inline router.py]: Looks good to me

File: router.py [hunk 6]
```diff
@@ -44,8 +55,43 @@ def receive_packet(self, packet):
...
+        # If we get a routing packet, we need to update the routing table
+        if isinstance(packet, RoutingPacket):
+            self.update_routing_table(packet, link_id)
+        else: # We need to send the packet to its dest host
+            self.send_packet(packet)
+
+    def send_routing_packets(self):
+        """Sends out the routing table to all of the router's links."""
+
+        for _, link in self.links:
```
Comment (hatooku) [inline router.py]: should be link in self.links.itervalues()

File: router.py [hunk 7]
```diff
@@ -44,8 +55,43 @@ def receive_packet(self, packet):
...
+        Args:
+            routing_packet (Packet): The routing packet we received.
+
+        """
+
+        # Check if we should update the routing table
+        changed = False
+        cost = self.links[link_id].prop_delay
+
+        for key, val in routing_packet.routing_table:
```
Comment (hatooku) [inline router.py]: 1. should use routing_table.items()
2. val is a tuple, not a number
3. should rename key and val to more descriptive variable names

File: router.py [hunk 8]
```diff
@@ -44,8 +55,43 @@ def receive_packet(self, packet):
...
+            if key not in self.routing_table or 
+            self.routing_table[key] + cost < val:
```
Comment (hatooku) [inline router.py]: should be self.routing_table[key][1]

File: router.py [hunk 9]
```diff
@@ -85,11 +85,12 @@ def update_routing_table(self, routing_packet, link_id):
         changed = False
         cost = self.links[link_id].prop_delay
 
-        for key, val in routing_packet.routing_table:
-            if key not in self.routing_table or 
-            self.routing_table[key] + cost < val:
+        for node_id, link_info in routing_packet.routing_table.iteritems():
+            if node_id not in self.routing_table or 
+            self.routing_table[node_id][1] + cost < link_info[1]:
```
Comment (hatooku) [inline router.py]: should switch self.routing_table[node_id][1] with link_info[1]

**Pull request merged**